website/util/send_digest.py

```diff
@@ -0,0 +1,162 @@
+"""Script for sending OSF email digests to subscribed users and removing the records once sent."""
+
+import datetime
+import urlparse
+import mock
+import unittest
+from bson.code import Code
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+from framework import sentry
+from framework.auth.core import User
+from framework.mongo import database as db
+from framework.tasks import app
+from framework.tasks.handlers import queued_task
+from website import mails, settings
+from website.app import init_app
+from website.util import web_url_for
+from website.notifications.model import DigestNotification
+from website.notifications.utils import NotificationsDict
+from tests.base import OsfTestCase
+from tests.factories import DigestNotificationFactory, UserFactory, ProjectFactory
+import nose.tools
+
+
+def main():
+    init_app(routes=False)
+    grouped_digests = group_digest_notifications_by_user()
+    print send_digest(grouped_digests)
+
+
+@queued_task
+@app.task
+def send_digest(grouped_digests):
+    for group in grouped_digests:
+        try:
+            user = User.load(group['user_id'])
+        except NoResultsFound:
+            sentry.log_exception()
+            sentry.log_message("A user with this username does not exist.")
+            user = None
+
+        info = group['info']
+        sorted_messages = group_messages(info)
+
+        if user and sorted_messages:
+            mails.send_mail(
+                to_addr=user.username,
+                mail=mails.DIGEST,
+                name=user.fullname,
+                message=sorted_messages,
+                url=urlparse.urljoin(settings.DOMAIN, 'settings/notifications/')
+            )
+
+    DigestNotification.remove(Q('timestamp', 'lt', datetime.datetime.utcnow()) &
+                              Q('timestamp', 'gte', datetime.datetime.utcnow() - datetime.timedelta(hours=24)))
+
+
+def group_messages(notifications):
+    d = NotificationsDict()
+    for n in notifications:
+        d.add_message(n['node_lineage'], n['message'])
+    return d
+
+
+def group_digest_notifications_by_user():
+    return db['digestnotification'].group(
+        key={'user_id': 1},
+        condition={'timestamp': {'$lt': datetime.datetime.utcnow(),
+                                 '$gte': datetime.datetime.utcnow() - datetime.timedelta(hours=24)}},
+        initial={'info': []},
+        reduce=Code("""function(curr, result) {
+                            info = {
+                                'message': {
+                                    'message': curr.message,
+                                    'timestamp': curr.timestamp
+                                },
+                                'node_lineage': curr.node_lineage
+                            }
+                            result.info.push(info);
+                    };
+                    """))
+
+
+class TestSendDigest(OsfTestCase):
```

Comment:
That is the convention for one-off scripts such as migrations. Since we will be using this script more-than-once, we will want to make sure that it runs with our CI builds.